flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -575,8 +576,51 @@ private void createManagedDirectory(Path managedPath) {
     @Override
     public void close() throws IOException {}
 
+    // ------------------------------------------------------------------------
+    //  restore
+    // ------------------------------------------------------------------------
+
+    @Override
+    public void restoreStateHandles(
+            long checkpointId, SubtaskKey subtaskKey, Stream<SegmentFileStateHandle> stateHandles) {
+
+        Set<LogicalFile> uploadedLogicalFiles;
+        synchronized (lock) {
+            uploadedLogicalFiles =
+                    uploadedStates.computeIfAbsent(checkpointId, id -> new HashSet<>());
+        }
+
+        stateHandles.forEach(
+                fileHandle -> {
+                    PhysicalFile physicalFile =
+                            new PhysicalFile(
+                                    null,
+                                    fileHandle.getFilePath(),
+                                    physicalFileDeleter,
```

Comment:
So for now will this `FileMergingSnapshotManager` delete underlying files after rescaling?

Reply:
IIUC, In the case of the job failover, need `FileMergingSnapshotManager` to delete the physical file; but in the case of the job rescaling, need jobManager to delete the physical file dir.

I will correct this logic. But i meet a problem where the TaskManager side has no way of knowing if the recovery is a rescaling situation (especially for the operator state rescaling).

Reply:
Can `StreamTaskStateInitializerImpl#keyGroupRange` and the `keyGroupRange` of one `KeyedStateHandle` be used as judgment conditions?

And for operator state rescaling, the new TMs do not claim ownership of these files, and they are always deleted by the JM.

Reply:
@fredia  In the case of the job restore (not failover and not rescaling),  the jobID may be changed, and I think the JM should claim ownership of these files to delete them.  In this case, however, the parallelism of the job has not changed, so the `keyGroupRange` of the `KeyedStateHandle` cannot be used as judgment conditions.

Reply:
As discussed offline with @Zakelly , we can distinguish whether the restored filePath is under the managedDir scope (`managedSharedStateDir` or `managedExclusiveStateDir`),   to determine whether the file should be deleted by the `FileMergingSnapshotManager`.